src/kv/raw_serialise.h

```diff
@@ -0,0 +1,200 @@
+      serialized::write(
+        data_,
+        size_,
+        reinterpret_cast<const uint8_t*>(entry.data()),
+        entry_size_bytes);
+    }
+
+    void serialise_string(const std::string& str)
+    {
+      size_t size_before = buf.size();
+      buf.resize(buf.size() + sizeof(size_t) + str.size());
+
+      auto data_ = buf.data() + size_before;
+      auto size_ = buf.size() - size_before;
+      serialized::write(data_, size_, str);
+    }
+
+  public:
+    RawWriter() = default;
+
+    template <typename T>
+    void append(const T& entry)
+    {
+      if constexpr (
+        nonstd::is_std_vector<T>::value ||
+        std::is_same_v<T, kv::serialisers::SerialisedEntry>)
+      {
+        serialise_entry(entry.size() * sizeof(typename T::value_type));
+        if (entry.size() > 0)
+        {
+          serialise_vector(entry);
+        }
+      }
+      else if constexpr (std::is_same_v<T, std::string>)
+      {
+        serialise_string(entry);
+      }
+      else if constexpr (std::is_integral_v<T>)
+      {
+        serialise_entry(entry);
+      }
+      else
+      {
+        static_assert(
+          nonstd::dependent_false<T>::value, "Can't serialise this type");
+      }
+    }
+
+    void clear()
+    {
+      buf.clear();
+    }
+
+    std::vector<uint8_t> get_raw_data()
+    {
+      return {buf.data(), buf.data() + buf.size()};
+    }
+  };
+
+  class RawReader
+  {
+  public:
+    const uint8_t* data_ptr;
+    size_t data_offset;
+    size_t data_size;
+
+    /** Reads the next entry, advancing data_offset
+     */
+    template <typename T>
+    T read_entry()
+    {
+      auto remainder = data_size - data_offset;
+      auto data = data_ptr + data_offset;
+      const auto entry = serialized::read<T>(data, remainder);
+      const auto bytes_read = data_size - data_offset - remainder;
+      data_offset += bytes_read;
+      return entry;
+    }
+
+    /** Reads the next size-prefixed entry
+     */
+    size_t read_size_prefixed_entry(size_t& start_offset)
+    {
+      auto remainder = data_size - data_offset;
+      auto entry_size = read_entry<size_t>();
+
+      if (remainder < entry_size)
+      {
+        throw std::runtime_error(fmt::format(
+          "Expected {} byte entry, found only {}", entry_size, remainder));
+      }
+
+      start_offset = data_offset;
+      data_offset += entry_size;
+
+      return entry_size;
+    }
+
+  public:
+    RawReader(const RawReader& other) = delete;
+    RawReader& operator=(const RawReader& other) = delete;
+
+    RawReader(const uint8_t* data_in_ptr = nullptr, size_t data_in_size = 0)
+    {
+      init(data_in_ptr, data_in_size);
+    }
+
+    void init(const uint8_t* data_in_ptr, size_t data_in_size)
+    {
+      data_offset = 0;
+      data_ptr = data_in_ptr;
+      data_size = data_in_size;
+    }
+
+    template <typename T>
+    T read_next()
+    {
+      if constexpr (
+        nonstd::is_std_vector<T>::value ||
```

Comment:
Do we ever actually use this for `std::vector<T>`, rather than just `std::vector<uint8_t>`? If we don't need arbitrary types, I think this and `std::string` can share more code and this can be simpler (no `sizeof` corrections, which I think is wrong for any struct with padding in it).

Reply:
We actually do for the serialisation of the view history in snapshots. 

Good point for the `sizeof` and padding! We could check that `T` is integral, or maybe use https://en.cppreference.com/w/cpp/types/has_unique_object_representations?

Reply:
After staring at this some more its actually safe - although we're writing directly to a `std::vector`'s `.data()`, we're only doing it with the contents we previously dumped _from_ a `std::vector`'s `.data()`, so it should be safe.